source/slang/slang-ir-diff-jvp.cpp

```diff
@@ -324,14 +392,81 @@ struct JVPDerivativeContext
         //
         SharedIRBuilder* sharedBuilder = &sharedBuilderStorage;
         sharedBuilder->init(module);
-
-        // Run through all the global-level instructions, 
-        // looking for callables.
-        // Note: We're only processing global callables (IRGlobalValueWithCode)
-        // for now.
-        // 
+    
         IRBuilder builderStorage(sharedBuilderStorage);
         IRBuilder* builder = &builderStorage;
+
+        // processMarkedGlobalFunctions(builder);
+        return processReferencedFunctions(builder);
+    }
+
+    IRInst* lookupJVPReference(IRInst* primalFunction)
+    {
+        if(auto jvpDefinition = primalFunction->findDecoration<IRJVPDerivativeReferenceDecoration>())
+            return jvpDefinition->getJVPFunc();
+        
+        return nullptr;
+    }
+
+    // Recursively process instructions looking for JVP calls (kIROp_JVPDifferentiate),
+    // then check that the referenced function is marked correctly for differentiation.
+    //
+    bool processReferencedFunctions(IRBuilder* builder)
+    {
+        IRWorkQueue* workQueue = &(workQueueStorage);
+
+        // Put the top-level inst into the queue.
+        workQueue->push(module->getModuleInst());
+        
+        // Keep processing items until the queue is complete.
+        while (IRInst* workItem = workQueue->pop())
+        {
+            for(auto child = workItem->getFirstChild(); child; child = child->getNextInst())
+            {
+                // Either the child instruction has more children (func/block etc..)
+                // and we add it to the work list for further processing, or 
+                // it's an ordinary inst in which case we check if it's a JVPDifferentiate
+                // instruction.
+                //
+                if (child->getFirstChild() != nullptr)
+                    workQueue->push(child);
+                
+                if (auto jvpDiffInst = as<IRJVPDifferentiate>(child))
+                {
+                    auto baseFunction = jvpDiffInst->getBaseFn();
+                    // If the JVP Reference already exists, no need to
+                    // differentiate again.
+                    //
+                    if(lookupJVPReference(baseFunction)) continue;
+
+                    if (isFunctionMarkedForJVP(as<IRGlobalValueWithCode>(baseFunction)))
+                    {
+                        IRFunc* jvpFunction = emitJVPFunction(builder, as<IRFunc>(baseFunction));
+                        builder->addJVPDerivativeReferenceDecoration(baseFunction, jvpFunction);
+                        workQueue->push(jvpFunction);
+                    }
+                    else
+                    {
+                        // TODO(Sai): This would probably be better with a more specific
+                        // error code.
+                        getSink()->diagnose(jvpDiffInst->sourceLoc,
```

Comment:
If we implement the checking in the front end then we should not reach the situation here. Whenever possible we should be doing semantic checks instead of reporting errors in an IR pass.